source/addonStore/dataManager.py

```diff
@@ -0,0 +1,227 @@
+	currentVersion = addonAPIVersion.CURRENT
+	year, major, minor = currentVersion
+	return f"{year}.{major}.{minor}"
+
+
+def _getAddonStoreURL(channel: Channel, lang: str, nvdaApiVersion: str) -> str:
+	_baseURL = "https://nvaccess.org/addonStore/"
+	return _baseURL + f"{lang}/{channel.value}/{nvdaApiVersion}.json"
+
+
+@dataclasses.dataclass
+class CachedAddonModel:
+	availableAddons: List[AddonDetailsModel]
+	cachedAt: datetime
+
+
+class AddonFileDownloader:
+	OnCompleteT = Callable[[AddonDetailsModel, Optional[os.PathLike]], None]
+
+	def __init__(self, cacheDir: os.PathLike):
+		self._cacheDir = cacheDir
+		self.progress: Dict[AddonDetailsModel, int] = {}  # Number of chunks received
+		self._pending: Dict[Future, Tuple[AddonDetailsModel, AddonFileDownloader.OnCompleteT]] = {}
+		self.complete: Dict[AddonDetailsModel, os.PathLike] = {}  # Path to downloaded file
+		self._executor = ThreadPoolExecutor(
+			max_workers=1,
+			thread_name_prefix="AddonDownloader",
+		)
+
+	def download(
+			self,
+			addonData: AddonDetailsModel,
+			onComplete: OnCompleteT
+	):
+		self.progress[addonData] = 0
+		f: Future = self._executor.submit(
+			self._download, addonData,
+		)
+		self._pending[f] = addonData, onComplete
+		f.add_done_callback(self._done)
+
+	def _done(self, downloadAddonFuture: Future):
+		isCancelled = downloadAddonFuture not in self._pending
+		addonId = "CANCELLED" if isCancelled else self._pending[downloadAddonFuture][0].addonId
+		log.debug(f"Done called for {addonId}")
+		if isCancelled:
+			log.debug("Download was cancelled, not calling onComplete")
+			return
+		if not downloadAddonFuture.done() or downloadAddonFuture.cancelled():
+			log.error("Logic error with download in BG thread.")
+			return
+		if downloadAddonFuture.exception():
+			log.error(f"Unhandled exception in _download", exc_info=downloadAddonFuture.exception())
+			return
+		addonData, onComplete = self._pending.pop(downloadAddonFuture)
+		del self.progress[addonData]
+		cacheFilePath: Optional[os.PathLike] = downloadAddonFuture.result()
+		self.complete[addonData] = cacheFilePath
+		onComplete(addonData, cacheFilePath)
+
+	def cancelAll(self):
+		log.debug("Cancelling all")
+		for f in self._pending.keys():
+			f.cancel()
+		self._executor.shutdown(wait=False)
+		self._executor = None
+		self.progress.clear()
+		self._pending.clear()
+
+	def _download(self, addonData: AddonDetailsModel) -> Optional[os.PathLike]:
+		log.debug(f"starting download: {addonData.addonId}")
+		cacheFilePath = os.path.join(
+			self._cacheDir,
+			self._getCacheFilenameForAddon(addonData)
+		)
+		inProgressFilePath = cacheFilePath + ".download"
+		if addonData not in self.progress:
+			log.debug("the download was cancelled before it started.")
+			return None  # The download was cancelled
+		try:
+			with requests.get(addonData.addonURL, stream=True) as r:
+				with open(inProgressFilePath, 'wb') as fd:
+					for chunk in r.iter_content(chunk_size=128000):

```

Comment:
Is this chunk size chosen arbitrarily?

Reply:
Yes, essentially.

Reply:
See https://stackoverflow.com/questions/23369625/ideal-chunk-size-for-python-requests

it looks like bigger chunk sizes are better for performance. Given we're dealing with add-ons that are usually smaller than 1MB, setting the chunk size to 1mb will probably mean that it's written to disk in just one write, wherea's 128 kb means that we might have multiple unnecessary writes.

Reply:
Smaller writes allows it to be interrupted, and provides a more responsive user experience. This is a tradeoff on throughput.
On very slow connections this would be important.

Reply:
This has been addressed with an in code comment ([d1677a4](https://github.com/nvaccess/nvda/pull/13985/commits/d1677a46410b432822a687099062d82b1f69eca9)). There is certainly room for more elaborate schemes.